### PR TITLE
Add API surface for messaging sub client

### PR DESCRIPTION
**Pull request by abhipsaMisra**
APIs for cloud to device telemetry related operations. One one set of APIs missing right now are the c2d send APIs. These APIs had previously been implemented on AMQP, and have not been implemented over Http in the current version of the swagger file. We have opted not to have the hand-written implementation over amqp as a part of the track 2 client library.

The md file can be viewed here: https://github.com/Azure/azure-sdk-for-net/blob/feature/iot/abmisr/messagingClient/sdk/iot/Azure.Iot.Hub.Service/src/API%20Design.md

### CODE REVIEW

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 1]
```diff
@@ -64,8 +64,43 @@ APIs for using IotHub v2 jobs
 
 <details><summary><b>Messages</b></summary>
 Feedback messages, sending cloud to device messages (missing from current swagger), and purging cloud to device message queue
-```csharp
 
+```csharp
+public class Messages
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:69]: Will this only be about CloudToDevice messages?

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 2]
```diff
@@ -64,8 +64,43 @@ APIs for using IotHub v2 jobs
...
+{
+    /// <summary>
+    /// Retrieve feedback notification for cloud to device messages.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>TODO: Swagger defines the response as the Http response returned from the service (and not the feedback message).</returns>
+    public virtual async Task<Response> GetCloudToDeviceMessageFeedbackAsync(CancellationToken cancellationToken = default) { }
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:76]: Including the "CloudToDeviceMessage" part makes each method a mouthful.

Also, I don't know what "Feedback" is supposed to mean. The method name as a whole is hard to understand.

Might not be much you can do about it, but perhaps we can brainstorm.

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md (hunk 1 above)
Comment (abhipsaMisra) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:69]: yes; should I rename the client to reflect that?

**Review by drwill-ms — APPROVED** [review]

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md (hunk 2 above)
Comment (abhipsaMisra) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:76]: I agree, the c2d message feedback flow is not very straight-forward and easy to use. We can go over the context of the flow in today's meeting and hopefully come up with a better name for this.

Comment (bikamani) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:76]: Is it better to call it **Receive**CloudToDeviceMessage**s**FeedbackAsync instead?

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md (hunk 1 above)
Comment (drwill-ms) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:69]: Something to consider

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md (hunk 2 above)
Comment (abhipsaMisra) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:76]: I can get on board with that; let's discuss this in today's meeting.

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 3]
```diff
@@ -64,8 +64,43 @@ APIs for using IotHub v2 jobs
...
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The Http response.</returns>
+    public virtual async Task<Response> AbandonCloudToDeviceMessageFeedbackAsync(string lockToken, CancellationToken cancellationToken = default) { }
+
+    /// <summary>
+    /// Purge the cloud to device message queue for a device.
+    /// </summary>
+    /// <param name="id">The unique identifier of the device.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The result of the cloud to device message queue purge operation.</returns>
```
Comment (bikamani) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:101]: Do we need word queue in summary/return? It makes it confusing in my opinion

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 4]
```diff
@@ -64,8 +64,43 @@ APIs for using IotHub v2 jobs
...
+    public virtual async Task<Response<PurgeMessageQueueResult>> PurgeCloudToDeviceMessageQueueAsync(string id, CancellationToken cancellationToken = default) { }
+}
```
Comment (bikamani) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:103]: Is this always for device? If so, we can call it deviceId

Comment (abhipsaMisra) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:103]: I am not sure if this always only for device, but yes, the `id` is the `deviceId`, so I will rename it. 
The `PurgeMessageQueueResult` type has a parameter for module Id.

**Comment by timtay-microsoft** [comment]
An interesting question I have for the service team around these APIs is why is there a purge message queue for devices, but not for modules? We can talk to them about that when we are reviewing these APIs with them

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 5]
```diff
@@ -64,8 +64,43 @@ APIs for using IotHub v2 jobs
...
-```csharp
 
+```csharp
+public class CloudToDeviceMessages
+{
+    /// <summary>
+    /// Retrieve feedback notification for cloud to device messages.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>TODO: Swagger defines the response as the Http response returned from the service (and not the feedback message).</returns>
```
Comment (timtay-microsoft) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:75]: Yikes. I can't wait to here the service explanation for this one.

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md (hunk 3 above)
Comment (timtay-microsoft) [reply sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:101]: Maybe just "The result of the purge operation" for simplicity?

**Comment by abhipsaMisra** [comment]
> An interesting question I have for the service team around these APIs is why is there a purge message queue for devices, but not for modules? We can talk to them about that when we are reviewing these APIs with them

I have the feeling the purge message queue purges the message queue for any module associated with the device? (I really hope this is not the case).
This is because the response type `PurgeMessageQueueResult` has a parameter for `moduleId`, even though we don't have an explicit API that takes in a module ID for the purge operation.

**Comment by abhipsaMisra** [comment]
Closing this, since the unavailability of service client c2d API over http renders all the other c2d APIs useless for now.

File: sdk/iot/Azure.Iot.Hub.Service/src/API Design.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+</details>
+
+<details><summary><b>Messages</b></summary>
+Feedback messages, sending cloud to device messages (missing from current swagger), and purging cloud to device message queue
+
+```csharp
+public class CloudToDeviceMessages
+{
+    /// <summary>
+    /// Retrieve feedback notification for cloud to device messages.
```
Comment (abhipsaMisra) [inline sdk/iot/Azure.Iot.Hub.Service/src/API Design.md:72]: why call it feedback? it seems to be the "result" of the message sent from service to device

**Pull request closed without merging**